src/main/java/io/wispforest/owo/serialization/Endec.java

```diff
@@ -0,0 +1,269 @@
+                StructField.of("k", keyEndec, Map.Entry::getKey),
+                StructField.of("V", valueEndec, Map.Entry::getValue),
+                Map::entry
+        );
+
+        return mapEntryEndec.list().then(entries -> {
+            Map<K, V> map = supplier.get();
+
+            for (Map.Entry<K, V> entry : entries) map.put(entry.getKey(), entry.getValue());
+
+            return map;
+        }, kvMap -> List.copyOf(kvMap.entrySet()));
+    }
+
+    static <T> AttributeEndecBuilder<T> ifAttr(Endec<T> endec, SerializationAttribute attribute){
+        return new AttributeEndecBuilder<>(endec, attribute);
+    }
+
+    //--
+
+    default ListEndec<T> list(){
+        return new ListEndec<>(this);
+    }
+
+    default MapEndec<String, T> map(){
+        return MapEndec.of(this);
+    }
+
+    default <R> Endec<R> then(Function<T, R> getter, Function<R, T> setter) {
+        return new Endec<>() {
+            @Override
+            public <E> void encode(Serializer<E> serializer, R value) {
+                Endec.this.encode(serializer, setter.apply(value));
+            }
+
+            @Override
+            public <E> R decode(Deserializer<E> deserializer) {
+                return getter.apply(Endec.this.decode(deserializer));
+            }
+        };
+    }
+
+    default Endec<Optional<T>> ofOptional(){
+        return new Endec<>() {
+            @Override
+            public <E> void encode(Serializer<E> serializer, Optional<T> value) {
+                serializer.writeOptional(Endec.this, value);
+            }
+
+            @Override
+            public <E> Optional<T> decode(Deserializer<E> deserializer) {
+                return deserializer.readOptional(Endec.this);
+            }
+        };
+    }
+
+    default Endec<@Nullable T> ofNullable(){
+        return ofOptional().then(o -> o.orElse(null), Optional::ofNullable);
+    }
+
+    default Endec<T> onError(TriConsumer<Serializer, T, Exception> encode, BiFunction<Deserializer, Exception, T> decode){
+        return new Endec<>() {
+            @Override
+            public <E> void encode(Serializer<E> serializer, T value) {
+                try {
+                    Endec.this.encode(serializer, value);
+                } catch (Exception e){
+                    encode.accept(serializer, value, e);
+                }
+            }
+
+            @Override
+            public <E> T decode(Deserializer<E> deserializer) {
+                try {
+                    return Endec.this.decode(deserializer);
+                } catch (Exception e) {
+                    return decode.apply(deserializer, e);
+                }
+            }
+        };
+    }
+
+    default Codec<T> codec(){
+        return new CooptCodec<>(this);
+    }
+
+    //--
+
+    <E> void encode(Serializer<E> serializer, T value);
+
+    <E> T decode(Deserializer<E> deserializer);
+
+    default <E> E encode(Supplier<Serializer<E>> serializerCreator, T value){
```

Comment:
why does this take a supplier if it just invokes it once anyways

Reply:
It is an easy way of constructing a supplier of the desired serializer type and returning the value held by such serializer as the return.